File: my-folder/0003-longest-substring-without-repeating-characters/solution.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        set<char> sub;
        int start = 0;
        int max_l = 0;

        for (int end=0;end<s.size();end++)
        {
            char a = s[end];

            auto itr = sub.find(a);
            if (itr!=sub.end())
            {
                
                while(s[start]!=a)
                {
                    sub.erase(s[start]);
                    start++;
                }
                sub.erase(s[start]);
                start++;
                
            }
            sub.insert(a);
            max_l = max(max_l, end-start+1);
        }
        return max_l;
    }
};

File: my-folder/0003-longest-substring-without-repeating-characters/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n = len(s)

        table = defaultdict(int)
        left = right = 0

        ans = 0
        while right < n:
            table[s[right]]+=1

            while left <= right and table[s[right]] > 1:
                table[s[left]]-=1
                if table[s[left]] == 0:
                    del table[s[left]]
                left+=1

            ans = max(ans, (right - left) +1)
            right+=1
        
        return ans







        
        
        
        
        
        
        
        
        
        
        
        
        










        
        
        
        
        
        # table = defaultdict(int)
        # n = len(s)
        # left = 0
        # right = 0

        # ans = 0
        # while right<n:
        #     table[s[right]]+=1

        #     while left<=right and table[s[right]]>1:
        #         table[s[left]]-=1
        #         if table[s[left]]==0:
        #             del table[s[left]]
        #         left+=1
        #     ans = max(ans,(right-left+1))
        #     right+=1
        # return ans
